src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,763 @@
+                        (k, false, true)
+                    }
+                    (Some(k), None) => {
+                        // Write cursor yields something but lock cursor yields `None`:
+                        // We need to further step write cursor to our desired version
+                        (Key::truncate_ts_for(k)?, true, false)
+                    }
+                    (Some(wk), Some(lk)) => {
+                        let write_user_key = Key::truncate_ts_for(wk)?;
+                        match write_user_key.cmp(lk) {
+                            Ordering::Less => {
+                                // Write cursor user key < lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to does not exist.
+                                (write_user_key, true, false)
+                            }
+                            Ordering::Greater => {
+                                // Write cursor user key > lock cursor, it means we got a lock of a
+                                // key that does not have a write. In SI, we need to check if the
+                                // lock will cause conflict.
+                                (lk, false, true)
+                            }
+                            Ordering::Equal => {
+                                // Write cursor user key == lock cursor, it means the lock of the
+                                // current key that write cursor is pointing to *exists*.
+                                (lk, true, true)
+                            }
+                        }
+                    }
+                };
+
+                // Use `from_encoded_slice` to reserve space for ts, so later we can append ts to
+                // the key or its clones without reallocation.
+                (Key::from_encoded_slice(res.0), res.1, res.2)
+            };
+
+            // `result` stores intermediate values, including KeyLocked errors (but not other kind
+            // of errors). If there is KeyLocked errors, we should be able to continue scanning.
+            let mut result = Ok(None);
+
+            // `get_ts` is the real used timestamp. If user specifies `MaxInt64` as the timestamp,
+            // we need to change it to a most recently available one.
+            let mut get_ts = self.ts;
+
+            // `met_next_user_key` stores whether the write cursor has been already pointing to
+            // the next user key. If so, we don't need to compare it again when trying to step
+            // to the next user key later.
+            let mut met_next_user_key = false;
+
+            if has_lock {
+                match self.isolation_level {
+                    IsolationLevel::SI => {
+                        // Only needs to check lock in SI
+                        match super::util::load_and_check_lock_from_cursor(
+                            &mut self.lock_cursor,
```

Comment:
Why not pass lock_cursor.value() directly?

Reply:
We have a similar function in the original refinement PR:

```
pub fn load_and_check_lock<S>(
    snapshot: &S,
    key: &Key,
    ts: u64,
    statistics: &mut Statistics,
) -> Result<u64>
where
    S: Snapshot,
{
    if let Some(lock) = load_lock(snapshot, key, statistics)? {
        return check_lock(key, ts, &lock);
    }
    Ok(ts)
}
```

so a new `load_and_check_lock_from_cursor` using exactly the same signature (but changes snapshot to a cursor) seems to fit better.

Reply:
Why we need to extract two simple code into such useless function here?
Another thing is this function is highly depended on the cursor which this is a implementation of Scanner::read_next. On the other hand, you introduce new return types for this function, it makes things complicated. The previous code is obvious and clear.

Reply:
New return type is not related to this extraction. The new return type is to solve the potential issue that errors other than KeyLocked are treated the same as KeyLocked, but in fact it shouldn't. KeyLocked is a kind of recoverable error, for which the caller can proceed scanning, while for other kind of Error they should be not recoverable.

Reply:
I can revert the function extraction, but I think it's better to keep the return type, just because previous return type is a union of non recoverable errors and recoverable errors and we processed it incorrectly.